### ISSUE TITLE
Generates duplicated Getters and Setters for some relations

### ISSUE THREAD

**Issue by brandomeniconi** [sequelize/sequelize-auto#466]
Hi,
I'm trying your tool and it works nicely, but has a tedious bug: in some models it generates duplicates getter and getters for each relation like so:

```TS
export class NewGift extends Model<NewGiftAttributes, NewGiftCreationAttributes> implements NewGiftAttributes {
  id!: number;
  treeId?: number;
  userId?: number;
  toUserId?: number;
  // [...]

  // NewGift belongsTo Tree
  getTree!: Sequelize.BelongsToGetAssociationMixin<Tree>;
  setTree!: Sequelize.BelongsToSetAssociationMixin<Tree, TreeId>;
  createTree!: Sequelize.BelongsToCreateAssociationMixin<Tree>;
  // NewGift belongsTo User
  getUser!: Sequelize.BelongsToGetAssociationMixin<User>;
  setUser!: Sequelize.BelongsToSetAssociationMixin<User, UserId>;
  createUser!: Sequelize.BelongsToCreateAssociationMixin<User>;
  // NewGift belongsTo User <-- duplicated
  getUser!: Sequelize.BelongsToGetAssociationMixin<User>;
  setUser!: Sequelize.BelongsToSetAssociationMixin<User, UserId>;
  createUser!: Sequelize.BelongsToCreateAssociationMixin<User>;
  // NewGift belongsTo Card
  getCard!: Sequelize.BelongsToGetAssociationMixin<Card>;
  setCard!: Sequelize.BelongsToSetAssociationMixin<Card, CardId>;
  createCard!: Sequelize.BelongsToCreateAssociationMixin<Card>;
  // NewGift hasOne Tree <-- duplicated
  getTree!: Sequelize.HasOneGetAssociationMixin<Tree>;
  setTree!: Sequelize.HasOneSetAssociationMixin<Tree, TreeId>;
  createTree!: Sequelize.HasOneCreateAssociationMixin<TreeCreationAttributes>;

  static initModel(sequelize: Sequelize.Sequelize): typeof NewGift {
  	NewGift.init({
  		id: {
  			autoIncrement: true,
  			type: DataTypes.INTEGER,
  			allowNull: false,
  			primaryKey: true
  		},
  		treeId: {
  			type: DataTypes.INTEGER,
  			allowNull: true,
  			references: {
  				model: "tree",
  				key: "id"
  			},
  			unique: "FK_XXXXXXXXXXXXX",
  			field: "tree_id"
  		},
  		userId: {
  			type: DataTypes.INTEGER,
  			allowNull: true,
  			references: {
  				model: "user",
  				key: "id"
  			},
  			field: "user_id"
  		},
  		toUserId: {
  			type: DataTypes.INTEGER,
  			allowNull: true,
  			references: {
  				model: "user",
  				key: "id"
  			},
  			field: "to_user_id"
  		},
                // [...]

```

there is a way to let sequelize-auto, use the field name and not the Model name to create g&s for relations?

Thanks!

**Comment by kippllo** [comment]
I can confirm this is happening for me too.

I am using:
- sequelize: 6.3.5
- sequelize-auto: 0.7.5,
- tedious: 8.3.0
- MS SQL Server: 2017

I'll try to post an example DB soon...

**Comment by kippllo** [comment]
Okay, I was able to duplicate the issue in this example DB:

### tblCustomers.sql:
```sql
SET ANSI_NULLS ON
GO

SET QUOTED_IDENTIFIER ON
GO

CREATE TABLE [dbo].[tblCustomers](
	[FacilityID] [int] NOT NULL,
	[CustomerID] [varchar](10) NOT NULL,
	[FNAME] [varchar](15) NOT NULL,
	[LNAME] [varchar](20) NOT NULL,
 CONSTRAINT [PK_tblCustomers] PRIMARY KEY CLUSTERED 
(
	[FacilityID] ASC,
	[CustomerID] ASC
)WITH (PAD_INDEX = OFF, STATISTICS_NORECOMPUTE = OFF, IGNORE_DUP_KEY = OFF, ALLOW_ROW_LOCKS = ON, ALLOW_PAGE_LOCKS = ON) ON [PRIMARY],
 CONSTRAINT [IX_tblCustomers] UNIQUE NONCLUSTERED 
(
	[FacilityID] ASC,
	[CustomerID] ASC
)WITH (PAD_INDEX = OFF, STATISTICS_NORECOMPUTE = OFF, IGNORE_DUP_KEY = OFF, ALLOW_ROW_LOCKS = ON, ALLOW_PAGE_LOCKS = ON) ON [PRIMARY]
) ON [PRIMARY]
GO
```

<br>

### tblTrips.sql:
```sql
SET ANSI_NULLS ON
GO

SET QUOTED_IDENTIFIER ON
GO

CREATE TABLE [dbo].[tblTrips](
	[FacilityID] [int] NOT NULL,
	[TripID] [int] IDENTITY(1,1) NOT NULL,
	[CustomerID] [varchar](10) NOT NULL,
	[Location] [varchar](100) NULL,
	[DateWhen] [datetime] NOT NULL,
 CONSTRAINT [PK_tblTrips] PRIMARY KEY CLUSTERED 
(
	[FacilityID] ASC,
	[TripID] ASC
)WITH (PAD_INDEX = OFF, STATISTICS_NORECOMPUTE = OFF, IGNORE_DUP_KEY = OFF, ALLOW_ROW_LOCKS = ON, ALLOW_PAGE_LOCKS = ON) ON [PRIMARY]
) ON [PRIMARY]
GO


ALTER TABLE [dbo].[tblTrips]  WITH NOCHECK ADD  CONSTRAINT [FK_tblTrips_tblCustomers] FOREIGN KEY([FacilityID], [CustomerID])
REFERENCES [dbo].[tblCustomers] ([FacilityID], [CustomerID])
ON UPDATE CASCADE
ON DELETE CASCADE
GO

ALTER TABLE [dbo].[tblTrips] CHECK CONSTRAINT [FK_tblTrips_tblCustomers]
GO
```

<br>

## I am using `sequelize-auto` in the "Programmatic API" form with the follow index.js file:
```js
const SequelizeAuto = require('sequelize-auto');
const SQLConn = require('./SQLConn.json');
const outputDir = './models';

var auto = new SequelizeAuto(SQLConn.database, SQLConn.username, SQLConn.password, {
    host: SQLConn.ip,
    port: SQLConn.port,
    dialect: 'mssql',
    dialectOptions:{
        options:{
            trustServerCertificate: true,
            encrypt: true,
            requestTimeout: 30000
        }
    },
    directory: outputDir,
    lang: 'ts',
    additional: {
        timestamps: false
    }
});

auto.run().then(data => {
    console.log('\nFinished successfully!');
});
```

<br>

## And it outputs these files:

### tblTrips.ts:
```ts
import Sequelize, { DataTypes, Model, Optional } from 'sequelize';
import type { tblCustomers, tblCustomersId } from './tblCustomers';

export interface tblTripsAttributes {
  FacilityID: number;
  TripID: number;
  CustomerID: string;
  Location?: string;
  DateWhen: Date;
}

export type tblTripsPk = "FacilityID" | "TripID";
export type tblTripsId = tblTrips[tblTripsPk];
export type tblTripsCreationAttributes = Optional<tblTripsAttributes, tblTripsPk>;

export class tblTrips extends Model<tblTripsAttributes, tblTripsCreationAttributes> implements tblTripsAttributes {
  FacilityID!: number;
  TripID!: number;
  CustomerID!: string;
  Location?: string;
  DateWhen!: Date;

  // tblTrips belongsTo tblCustomers
  gettblCustomer!: Sequelize.BelongsToGetAssociationMixin<tblCustomers>;
  settblCustomer!: Sequelize.BelongsToSetAssociationMixin<tblCustomers, tblCustomersId>;
  createtblCustomer!: Sequelize.BelongsToCreateAssociationMixin<tblCustomers>;
  // tblTrips belongsTo tblCustomers <-- duplicated
  gettblCustomer!: Sequelize.BelongsToGetAssociationMixin<tblCustomers>;
  settblCustomer!: Sequelize.BelongsToSetAssociationMixin<tblCustomers, tblCustomersId>;
  createtblCustomer!: Sequelize.BelongsToCreateAssociationMixin<tblCustomers>;

  static initModel(sequelize: Sequelize.Sequelize): typeof tblTrips {
    tblTrips.init({
    FacilityID: {
      type: DataTypes.INTEGER,
      allowNull: false,
      primaryKey: true,
      references: {
        model: 'tblCustomers',
        key: 'FacilityID'
      }
    },
    TripID: {
      autoIncrement: true,
      type: DataTypes.INTEGER,
      allowNull: false,
      primaryKey: true
    },
    CustomerID: {
      type: DataTypes.STRING(10),
      allowNull: false,
      references: {
        model: 'tblCustomers',
        key: 'FacilityID'
      }
    },
    Location: {
      type: DataTypes.STRING(100),
      allowNull: true
    },
    DateWhen: {
      type: DataTypes.DATE,
      allowNull: false
    }
  }, {
    sequelize,
    tableName: 'tblTrips',
    schema: 'dbo',
    timestamps: false,
    indexes: [
      {
        name: "PK_tblTrips",
        unique: true,
        fields: [
          { name: "FacilityID" },
          { name: "TripID" },
        ]
      },
    ]
  });
  return tblTrips;
  }
}
```

<br>

### tblCustomers.ts:
```ts
import Sequelize, { DataTypes, Model, Optional } from 'sequelize';
import type { tblTrips, tblTripsId } from './tblTrips';

export interface tblCustomersAttributes {
  FacilityID: number;
  CustomerID: string;
  FNAME: string;
  LNAME: string;
}

export type tblCustomersPk = "FacilityID" | "CustomerID";
export type tblCustomersId = tblCustomers[tblCustomersPk];
export type tblCustomersCreationAttributes = Optional<tblCustomersAttributes, tblCustomersPk>;

export class tblCustomers extends Model<tblCustomersAttributes, tblCustomersCreationAttributes> implements tblCustomersAttributes {
  FacilityID!: number;
  CustomerID!: string;
  FNAME!: string;
  LNAME!: string;

  // tblCustomers hasMany tblTrips
  gettblTrips!: Sequelize.HasManyGetAssociationsMixin<tblTrips>;
  settblTrips!: Sequelize.HasManySetAssociationsMixin<tblTrips, tblTripsId>;
  addtblTrips!: Sequelize.HasManyAddAssociationsMixin<tblTrips, tblTripsId>;
  removetblTrips!: Sequelize.HasManyRemoveAssociationsMixin<tblTrips, tblTripsId>;
  hastblTrips!: Sequelize.HasManyHasAssociationsMixin<tblTrips, tblTripsId>;
  counttblTrips!: Sequelize.HasManyCountAssociationsMixin;
  // tblCustomers hasMany tblTrips <-- duplicated
  gettblTrips!: Sequelize.HasManyGetAssociationsMixin<tblTrips>;
  settblTrips!: Sequelize.HasManySetAssociationsMixin<tblTrips, tblTripsId>;
  addtblTrips!: Sequelize.HasManyAddAssociationsMixin<tblTrips, tblTripsId>;
  removetblTrips!: Sequelize.HasManyRemoveAssociationsMixin<tblTrips, tblTripsId>;
  hastblTrips!: Sequelize.HasManyHasAssociationsMixin<tblTrips, tblTripsId>;
  counttblTrips!: Sequelize.HasManyCountAssociationsMixin;

  static initModel(sequelize: Sequelize.Sequelize): typeof tblCustomers {
    tblCustomers.init({
    FacilityID: {
      type: DataTypes.INTEGER,
      allowNull: false,
      primaryKey: true,
      unique: "IX_tblCustomers"
    },
    CustomerID: {
      type: DataTypes.STRING(10),
      allowNull: false,
      primaryKey: true,
      unique: "IX_tblCustomers"
    },
    FNAME: {
      type: DataTypes.STRING(15),
      allowNull: false
    },
    LNAME: {
      type: DataTypes.STRING(20),
      allowNull: false
    }
  }, {
    sequelize,
    tableName: 'tblCustomers',
    schema: 'dbo',
    timestamps: false,
    indexes: [
      {
        name: "IX_tblCustomers",
        unique: true,
        fields: [
          { name: "FacilityID" },
          { name: "CustomerID" },
        ]
      },
      {
        name: "PK_tblCustomers",
        unique: true,
        fields: [
          { name: "FacilityID" },
          { name: "CustomerID" },
        ]
      },
    ]
  });
  return tblCustomers;
  }
}
```

<br>

### init-models.ts:
```ts
import type { Sequelize, Model } from "sequelize";
import { tblCustomers } from "./tblCustomers";
import type { tblCustomersAttributes, tblCustomersCreationAttributes } from "./tblCustomers";
import { tblTrips } from "./tblTrips";
import type { tblTripsAttributes, tblTripsCreationAttributes } from "./tblTrips";

export {
  tblCustomers,
  tblTrips,
};

export type {
  tblCustomersAttributes,
  tblCustomersCreationAttributes,
  tblTripsAttributes,
  tblTripsCreationAttributes,
};

export function initModels(sequelize: Sequelize) {
  tblCustomers.initModel(sequelize);
  tblTrips.initModel(sequelize);

  tblTrips.belongsTo(tblCustomers, { foreignKey: "CustomerID"});
  tblCustomers.hasMany(tblTrips, { foreignKey: "CustomerID"});
  tblTrips.belongsTo(tblCustomers, { foreignKey: "FacilityID"});
  tblCustomers.hasMany(tblTrips, { foreignKey: "FacilityID"});

  return {
    tblCustomers: tblCustomers,
    tblTrips: tblTrips,
  };
}
```

## Final Thoughts:
I believe this problem happens because sequelize-auto is making **two different relationships for a composite foreign key**.

<br>

I hope this info helps!
Let me know if there is anything else I can do.

**Comment by steveschmitt** [comment]
Wow, thanks for the great repro!  I'll take a look.

**Comment by brandomeniconi** [comment]
Thanks @kippllo! 
Looking at the source code I think that `sequelize-auto` uses the related Model name to generate the relation functions names. If we have multiple relations to the same Model in the same schema, like:

```
Person
   id
   name

Car
   driverId  --> Person(id)
   passengerId --> Person(id)
```

in the `Car` model it will name both relations `getPerson` instead of `getDriver` and `getPassenger`.

@steveschmitt For us it will be great to name relation functions after the "field name", without the trailing `id` (if present). Do you think that's possible?

Thanks!

**Comment by kippllo** [comment]
@steveschmitt @brandomeniconi 
So, it seems there is two different cases that can cause this issue?
- Two different foreign keys pointing to two different records in the same table?
- Any composite foreign key?

**Comment by steveschmitt** [comment]
@kippllo  The problem with composite foreign keys is that [Sequelize doesn't really support that](https://github.com/sequelize/sequelize/issues/311).  

I can ensure that the **get/set/add/remove** association methods are added only once, but I don't know if the resulting code will actually work at runtime.  

Can you please edit your generated code, removing the duplicated lines, and see how it works at runtime?  Make sure it loads the correct related entities (matching **both** foreign keys).


@brandomeniconi For multiple relations to the same model, the associations need to be [aliased](https://sequelize.org/master/manual/assocs.html#multiple-associations-involving-the-same-models).  

I can attempt to generate the alias based on the property name.

**Comment by steveschmitt** [comment]
Fixed in 0.7.8

Now associations use foreign-key property names instead of model names
